python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -179,18 +183,32 @@ def get_nodes_to_launch(
             return self._legacy_worker_node_to_launch(
                 nodes, launching_nodes, node_resources,
                 resource_demands + request_resources_demands)
-        placement_group_nodes_to_add, node_resources, node_type_counts = \
+
+        spread_pg_nodes_to_add, node_resources, node_type_counts = \
             self.reserve_and_allocate_spread(
                 strict_spreads, node_resources, node_type_counts)
 
+        # Calculate the nodes to add for bypassing max launch limit for
+        # placement groups and spreads.
+        unfulfilled_placement_groups_demands, _ = get_bin_pack_residual(
+            node_resources, placement_group_demand_vector)
+        # Add 1 to account for the head node.
+        max_to_add = self.max_workers + 1 - sum(node_type_counts.values())
+        pg_demands_nodes_max_launch_limit = get_nodes_for(
+            self.node_types, node_type_counts, self.head_node_type, max_to_add,
+            unfulfilled_placement_groups_demands)
+        placement_groups_nodes_max_limit = {
+            node_type: spread_pg_nodes_to_add.get(node_type, 0) +
+            pg_demands_nodes_max_launch_limit.get(node_type, 0)
```

Comment:
Why do we add them together here? Shouldn't it be a `max()`?

Reply:
See comment above. this is why I split it to spread_pg_nodes_to_add and pg_demands_nodes_max_launch_limit that should be added. The first counts the spreads and the second counts the demands.